File: solutions/prolog/triangle/1/triangle.pl
valid(Side1, Side2, Side3) :-
  Side1 > 0,
  Side2 > 0,
  Side3 > 0,
  Side1 + Side2 >= Side3,
  Side1 + Side3 >= Side2,
  Side2 + Side3 >= Side1.

triangle(Side1, Side2, Side3, "equilateral") :-
  valid(Side1, Side2, Side3),
  Side1 =:= Side2,
  Side2 =:= Side3.

triangle(Side1, Side2, Side3, "isosceles") :-
  valid(Side1, Side2, Side3),
  Side1 =:= Side2.

triangle(Side1, Side2, Side3, "isosceles") :-
  valid(Side1, Side2, Side3),
  Side1 =:= Side3.

triangle(Side1, Side2, Side3, "isosceles") :-
  valid(Side1, Side2, Side3),
  Side2 =:= Side3.

triangle(Side1, Side2, Side3, "scalene") :-
  valid(Side1, Side2, Side3),
  Side1 =\= Side2,
  Side1 =\= Side3,
  Side2 =\= Side3.

File: solutions/erlang/triangle/1/src/triangle.erl
-module(triangle).
-export([kind/3]).

kind(A, B, C) when A*B*C == 0
  -> {error, "all side lengths must be positive"};
kind(A, B, C) when A + B =< C; A + C =< B; B + C =< A
  -> {error, "side lengths violate triangle inequality"};
kind(A, A, A) -> equilateral;
kind(A, B, C) when A == B; A == C; B == C -> isosceles;
kind(_,_,_) -> scalene.
